### PR TITLE
Predicate Server Access Integration

### PR DESCRIPTION
**Pull request by xacrimon**
This PR contains changes and primitives needed for integration with Server Access.
As of this PR, integration with server access is demoable.

This PR implements support for:
- node session create
- node session join
- node visibility
- general resource rbac

Paired with https://github.com/gravitational/predicate-lang/pull/24

### CODE REVIEW

**Comment by xacrimon** [comment]
Update: waiting on completion of https://github.com/gravitational/predicate-lang/issues/31 to rebase on the result of that

File: lib/auth/predicate/partial/builtin.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		ctx.If(
+			element.Length().Eq(zero),
+			element.AsAST(),
+			charUpper(element.Substring(zero, one)).Concat(rem).AsAST(),
+		))
+
+	return fnUpper, nil
+}
+
+//z3.RecAddDefinition(
```
Comment (nklaassen) [inline lib/auth/predicate/partial/builtin.go:51]: do we want to merge this commented out code?

File: lib/services/role.go [hunk 2]
```diff
@@ -2394,10 +2408,10 @@ func (p boolParser) Parse(string) (interface{}, error) {
 // CheckAccessToRule checks if the RoleSet provides access in the given
 // namespace to the specified resource and verb.
 // silent controls whether the access violations are logged.
-func (set RoleSet) CheckAccessToRule(ctx RuleContext, namespace string, resource string, verb string, silent bool) error {
+func (set RoleSet) CheckAccessToRule(ctx RuleContext, namespace string, resource string, verb string, silent bool) (predicateEngine.AccessDecision, error) {
 	whereParser, err := NewWhereParser(ctx)
 	if err != nil {
-		return trace.Wrap(err)
+		return predicateEngine.AccessUndecided, trace.Wrap(err)
```
Comment (nklaassen) [inline lib/services/role.go:2414]: why undecided when there is a parse error? should we play it safe and deny access unless we know for sure it is meant to be undecided?

File: lib/services/role.go [hunk 3]
```diff
@@ -2035,25 +2048,25 @@ func (set RoleSet) checkAccess(r AccessCheckable, mfa AccessMFAParams, matchers
...
 		matchLabels, labelsMessage, err := MatchLabels(getRoleLabels(role, types.Deny), allLabels)
 		if err != nil {
-			return trace.Wrap(err)
+			return predicateEngine.AccessUndecided, trace.Wrap(err)
 		}
 		if matchLabels {
 			debugf("Access to %v %q denied, deny rule in role %q matched; match(namespace=%v, label=%v)",
 				r.GetKind(), r.GetName(), role.GetName(), namespaceMessage, labelsMessage)
-			return trace.AccessDenied("access to %v denied. User does not have permissions. %v",
+			return predicateEngine.AccessUndecided, trace.AccessDenied("access to %v denied. User does not have permissions. %v",
```
Comment (nklaassen) [inline lib/services/role.go:2056]: isn't this an explicit deny? why does it return undecided?

**Review by nklaassen — COMMENTED** [review]
@xacrimon are you looking for a full review of this now, so that the entire diff on the `dev/predicate` branch can be merged to `master` in one big PR later on without too much scrutiny, or do you plan to eventually start merging smaller chunks to master with another review at that time?

If this were merging to master I would suggest breaking it up. There are type renames, refactors, changes to user certificates, and pretty fundamental changes to our RBAC checks here. Some of these deserve a pretty close eye. I think it would be much easier for reviewers (and you would have a much easier time merging these) if the PRs were broken down into logical chunks.

**Comment by xacrimon** [comment]
@nklaassen I'm expecting this sort of a review stage to be a good opportunity to point out issues which you may see coming to be problematic in the future and other things you notice that don't seem right. This is still very much a stage of stringing things together and seeing if they work. When we go to merge this to master, we'll split the changes up into incremental bits organized by section of codebase to make this easier to review in detail and point out subtle faults.

File: lib/auth/predicate/partial/builtin.go (hunk 1 above)
Comment (xacrimon) [reply lib/auth/predicate/partial/builtin.go:51]: sorry, this was reference code from z3Py I used. Removed

File: lib/services/role.go (hunk 2 above)
Comment (xacrimon) [reply lib/services/role.go:2414]: Throughout this code, undecided has been used as a sort of default value. This still errors, though, so it'll never return allow and no error at the top level. I think undecided better reflects how we'd reason about this mentally but it largely doesn't matter functionally here as the error ensures you never get that higher level access decision.

File: lib/services/role.go (hunk 3 above)
Comment (xacrimon) [reply lib/services/role.go:2056]: Good catch, thanks

**Comment by xacrimon** [comment]
On second thought @nklaassen. I'll split this

File: lib/services/role.go (hunk 2 above)
Comment (nklaassen) [reply lib/services/role.go:2415]: ah, somehow I didn't consider the error return, that's fine then

**Comment by nklaassen** [comment]
> All changes made on dev-predicate will be split into logical chunks for a detailed review and to make a security review later when we decide to join this feature into master. I this stage of the development process represents a good time to give design decision feedback and to point out things one believes may become a problem down the line, and to keep people in sync about what's going on. 

@xacrimon that's very good to know, thanks

**Pull request closed without merging**